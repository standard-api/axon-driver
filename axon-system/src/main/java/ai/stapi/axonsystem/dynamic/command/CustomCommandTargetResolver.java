package ai.stapi.axonsystem.dynamic.command;

import ai.stapi.graphsystem.messaging.command.AbstractCommand;
import org.axonframework.commandhandling.CommandMessage;
import org.axonframework.modelling.command.AnnotationCommandTargetResolver;
import org.axonframework.modelling.command.CommandTargetResolver;
import org.axonframework.modelling.command.VersionedAggregateIdentifier;
import org.jetbrains.annotations.NotNull;

public class CustomCommandTargetResolver implements CommandTargetResolver {
  
  private final AnnotationCommandTargetResolver annotationCommandTargetResolver;

  public CustomCommandTargetResolver() {
    this.annotationCommandTargetResolver = AnnotationCommandTargetResolver.builder().build();
  }

  @Override
  public VersionedAggregateIdentifier resolveTarget(@NotNull CommandMessage<?> command) {
    var payload = command.getPayload();
    if (payload instanceof AbstractCommand<?> abstractCommand) {
      return new VersionedAggregateIdentifier(
          abstractCommand.getTargetIdentifier(),
          null
      );
    }
    return this.annotationCommandTargetResolver.resolveTarget(command);
  }
}
