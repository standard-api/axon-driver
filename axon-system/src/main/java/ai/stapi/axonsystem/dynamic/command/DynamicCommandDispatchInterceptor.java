package ai.stapi.axonsystem.dynamic.command;

import ai.stapi.graphsystem.messaging.command.DynamicCommand;
import java.util.List;
import java.util.function.BiFunction;
import org.axonframework.commandhandling.CommandMessage;
import org.axonframework.commandhandling.GenericCommandMessage;
import org.axonframework.messaging.GenericMessage;
import org.axonframework.messaging.MessageDispatchInterceptor;
import org.jetbrains.annotations.NotNull;

public class DynamicCommandDispatchInterceptor  implements MessageDispatchInterceptor<CommandMessage<?>> {

  @NotNull
  @Override
  public BiFunction<Integer, CommandMessage<?>, CommandMessage<?>> handle(
      @NotNull List<? extends CommandMessage<?>> messages
  ) {
    return (integer, commandMessage) -> {
      if (commandMessage.getPayload() instanceof DynamicCommand dynamicCommand) {
        return new GenericCommandMessage<>(
            new GenericMessage<>(dynamicCommand),
            dynamicCommand.getSerializationType()
        );
      }
      return commandMessage;
    };
  }
}
