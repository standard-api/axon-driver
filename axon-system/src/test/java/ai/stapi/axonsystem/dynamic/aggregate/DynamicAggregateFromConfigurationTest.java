package ai.stapi.axonsystem.dynamic.aggregate;

import ai.stapi.axonsystem.dynamic.aggregate.config.DynamicAggregateFromConfigurationTestConfiguration;
import ai.stapi.axonsystem.dynamic.aggregate.testImplementations.ExampleDynamicAggregateCreatedFactory.ExampleDynamicAggregateCreated;
import ai.stapi.axonsystem.dynamic.aggregate.testImplementations.ExampleDynamicAggregateTypeConstructorCommandOgmProvider;
import ai.stapi.axonsystem.dynamic.aggregate.testImplementations.ExampleDynamicAggregateTypeCreateIfMissingCommandOgmProvider;
import ai.stapi.axonsystem.dynamic.aggregate.testImplementations.ExampleDynamicAggregateTypeMethodCommandOgmProvider;
import ai.stapi.graphsystem.messaging.command.DynamicCommand;
import ai.stapi.identity.UniqueIdentifier;
import ai.stapi.test.DomainTestCase;
import ai.stapi.test.schemaintegration.StructureDefinitionScope;
import java.util.Map;
import org.junit.jupiter.api.Test;
import org.springframework.context.annotation.Import;

@Import(DynamicAggregateFromConfigurationTestConfiguration.class)
@StructureDefinitionScope(DynamicAggregateDefinitionsLoader.SCOPE)
class DynamicAggregateFromConfigurationTest extends DomainTestCase {

  @Test
  void itCanCreateDynamicAggregate() {
    this.whenCommandIsDispatched(
        new DynamicCommand(
            new UniqueIdentifier("ExampleId"),
            ExampleDynamicAggregateTypeConstructorCommandOgmProvider.COMMAND_NAME,
            Map.of(
                "exampleConstructorAttribute", "Example Constructor Attribute Value"
            )
        )
    );
    this.thenLastEventOfTypeGraphApproved(
        ExampleDynamicAggregateCreated.class);

  }

  @Test
  void itCanCreateDynamicAggregateIfMissing() {
    this.whenCommandIsDispatched(
        new DynamicCommand(
            new UniqueIdentifier("ExampleId"),
            ExampleDynamicAggregateTypeCreateIfMissingCommandOgmProvider.COMMAND_NAME,
            Map.of(
                "exampleCreateIfMissingAttribute", "Example If Missing Attribute Value"
            )
        )
    );
    this.thenLastEventOfTypeGraphApproved(
        ExampleDynamicAggregateCreated.class);

  }

  @Test
  void itCanChangeDynamicAggregateByIfMissingIfAlreadyThere() {
    this.givenCommandIsDispatched(
        new DynamicCommand(
            new UniqueIdentifier("ExampleId"),
            ExampleDynamicAggregateTypeCreateIfMissingCommandOgmProvider.COMMAND_NAME,
            Map.of(
                "exampleCreateIfMissingAttribute", "Example If Missing Attribute Value"
            )
        )
    );
    this.whenCommandIsDispatched(
        new DynamicCommand(
            new UniqueIdentifier("ExampleId"),
            ExampleDynamicAggregateTypeCreateIfMissingCommandOgmProvider.COMMAND_NAME,
            Map.of(
                "exampleCreateIfMissingAttribute", "Example Changed If Missing Attribute Value"
            )
        )
    );
    this.thenLastEventOfTypeGraphApproved(
        ExampleDynamicAggregateCreated.class,
        2
    );
  }

  @Test
  void itCanChangeDynamicAggregate() {
    this.givenCommandIsDispatched(
        new DynamicCommand(
            new UniqueIdentifier("ExampleId"),
            ExampleDynamicAggregateTypeConstructorCommandOgmProvider.COMMAND_NAME,
            Map.of(
                "exampleConstructorAttribute", "Example Constructor Attribute Value"
            )
        )
    );
    this.whenCommandIsDispatched(
        new DynamicCommand(
            new UniqueIdentifier("ExampleId"),
            ExampleDynamicAggregateTypeMethodCommandOgmProvider.COMMAND_NAME,
            Map.of(
                "exampleMethodAttribute", "Example Method Attribute Value"
            )
        )
    );
    this.thenMergedGraphOfAggregateApproved("ExampleId");
  }

  @Test
  void itCanCreateTwoAggregatesOfSameType() {
    this.whenCommandIsDispatched(
        new DynamicCommand(
            new UniqueIdentifier("ExampleId"),
            ExampleDynamicAggregateTypeConstructorCommandOgmProvider.COMMAND_NAME,
            Map.of(
                "exampleConstructorAttribute", "Example Constructor Attribute Value"
            )
        )
    );
    this.whenCommandIsDispatched(
        new DynamicCommand(
            new UniqueIdentifier("ExampleId2"),
            ExampleDynamicAggregateTypeConstructorCommandOgmProvider.COMMAND_NAME,
            Map.of(
                "exampleConstructorAttribute", "Example Constructor Attribute Value 2"
            )
        )
    );
    this.whenCommandIsDispatched(
        new DynamicCommand(
            new UniqueIdentifier("ExampleId"),
            ExampleDynamicAggregateTypeMethodCommandOgmProvider.COMMAND_NAME,
            Map.of(
                "exampleMethodAttribute", "Example Method Attribute Value"
            )
        )
    );
    this.whenCommandIsDispatched(
        new DynamicCommand(
            new UniqueIdentifier("ExampleId2"),
            ExampleDynamicAggregateTypeMethodCommandOgmProvider.COMMAND_NAME,
            Map.of(
                "exampleMethodAttribute", "Example Method Attribute Value 2"
            )
        )
    );
    this.thenMergedGraphOfAggregateApproved("ExampleId");
  }
}