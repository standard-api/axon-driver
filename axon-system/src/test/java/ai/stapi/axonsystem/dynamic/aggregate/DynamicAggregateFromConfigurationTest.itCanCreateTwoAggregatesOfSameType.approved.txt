nodes:
	node_type: ExampleDynamicAggregateType
	node_hash: 00AD4
	node_edges:
	attributes:
		exampleConstructorAttribute -> Example Constructor Attribute Value
		exampleMethodAttribute -> Example Method Attribute Value

edges:
